### PR TITLE
[llvm] Ensure that soft float targets don't emit `fma()` libcalls.

### PR DESCRIPTION
**Pull request by alexrp**
The previous behavior could be harmful in some edge cases, such as emitting a call to `fma()` in the `fma()` implementation itself.

Do this by just being more accurate in `isFMAFasterThanFMulAndFAdd()`. This was already done for PowerPC; this commit just extends that to Arm, z/Arch, and x86. MIPS and SPARC already got it right, but I added tests for them too, for good measure.

Note: I don't have commit access.

### CODE REVIEW

**Comment by efriedma-quic** [comment]
Can you update the description in TargetLowering.h to mention this restriction?

**Comment by alexrp** [comment]
Done.

File: llvm/test/CodeGen/ARM/fmuladd-soft-float.ll [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+; SOFT-FLOAT-VFP64:       @ %bb.0:
+; SOFT-FLOAT-VFP64-NEXT:    push {r11, lr}
+; SOFT-FLOAT-VFP64-NEXT:    bl __muldf3
+; SOFT-FLOAT-VFP64-NEXT:    ldr r2, [sp, #8]
+; SOFT-FLOAT-VFP64-NEXT:    ldr r3, [sp, #12]
+; SOFT-FLOAT-VFP64-NEXT:    bl __adddf3
+; SOFT-FLOAT-VFP64-NEXT:    pop {r11, lr}
+; SOFT-FLOAT-VFP64-NEXT:    mov pc, lr
+    %1 = call double @llvm.fmuladd.f64(double %a, double %b, double %c)
+	ret double %1
```
Comment (arsenm) [inline llvm/test/CodeGen/ARM/fmuladd-soft-float.ll:71]: Weird indentation (also should use named values)

File: llvm/test/CodeGen/ARM/fmuladd-soft-float.ll [hunk 2]
```diff
@@ -0,0 +1,75 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=arm < %s | FileCheck %s -check-prefix=SOFT-FLOAT
+; RUN: llc -mtriple=arm -mattr=+vfp4d16sp < %s | FileCheck %s -check-prefix=SOFT-FLOAT-VFP32
+; RUN: llc -mtriple=arm -mattr=+vfp4d16sp,+fp64 < %s | FileCheck %s -check-prefix=SOFT-FLOAT-VFP64
+
+define float @fma_f32(float %a, float %b, float %c) "use-soft-float"="true" {
```
Comment (arsenm) [inline llvm/test/CodeGen/ARM/fmuladd-soft-float.ll:6]: Test name is misleading, it's not using FMA directly. Can you also add tests with the fmul contract + fadd contract pattern?

Also use attribute group?

File: llvm/test/CodeGen/ARM/fmuladd-soft-float.ll (hunk 1 above)
Comment (alexrp) [reply llvm/test/CodeGen/ARM/fmuladd-soft-float.ll:71]: Would you mind elaborating on the named values point? I'm fairly new to contributing to LLVM, so I don't know if there's some significance there that I'm missing. Even if I change it to a named value, I don't know that I can come up with anything better than `result` (or something equally generic). :thinking:

**Comment by alexrp** [comment]
Pushed updated tests.

**Comment by alexrp** [comment]
ping 🙂

File: llvm/test/CodeGen/ARM/fmuladd-soft-float.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/CodeGen/ARM/fmuladd-soft-float.ll]: The names don't mean anything, but anonymous values sometimes make it more difficult to update tests in the future 

https://llvm.org/docs/TestingGuide.html#best-practices-for-regression-tests

**Review by arsenm — COMMENTED** [review]
This doesn't feel like the right place for this, I would expect there to be some legality check that would fail. But there's already precedent for directly handling it here. Anyway basically lgtm

File: llvm/test/CodeGen/SystemZ/fmuladd-soft-float.ll [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+; SOFT-FLOAT-FPC-FAST-NEXT:    brasl %r14, __muldf3@PLT
+; SOFT-FLOAT-FPC-FAST-NEXT:    lgr %r3, %r13
+; SOFT-FLOAT-FPC-FAST-NEXT:    brasl %r14, __adddf3@PLT
+; SOFT-FLOAT-FPC-FAST-NEXT:    lmg %r13, %r15, 264(%r15)
+; SOFT-FLOAT-FPC-FAST-NEXT:    br %r14
+    %product = fmul double %a, %b
+    %result = fadd double %product, %c
+    ret double %result
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/SystemZ/fmuladd-soft-float.ll:221]: Should test some vector cases too

File: llvm/test/CodeGen/ARM/fmuladd-soft-float.ll [hunk 4]
```diff
@@ -0,0 +1,393 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=arm --fp-contract off < %s | FileCheck %s -check-prefix=SOFT-FLOAT-FPC-OFF
+; RUN: llc -mtriple=arm --fp-contract on < %s | FileCheck %s -check-prefix=SOFT-FLOAT-FPC-ON
+; RUN: llc -mtriple=arm --fp-contract fast < %s | FileCheck %s -check-prefix=SOFT-FLOAT-FPC-FAST
+; RUN: llc -mtriple=arm -mattr=+vfp4d16sp --fp-contract off < %s | FileCheck %s -check-prefix=SOFT-FLOAT-VFP32-FPC-OFF
```
Comment (arsenm) [inline llvm/test/CodeGen/ARM/fmuladd-soft-float.ll:5]: Avoid adding new uses of the fp-contract flag, and prefer explicit contract flags in the IR

**Comment by alexrp** [comment]
> This doesn't feel like the right place for this, I would expect there to be some legality check that would fail.

For what it's worth, what happened for the affected targets was that the `fmuladd` intrinsic was transformed into an FMA node in the SDAG due to `isFMAFasterThanFMulAndFAdd()` saying it's worthwhile to do so, and then legalization turned that FMA node into an `fma()` libcall due to soft float being in effect. That all seems sane enough to me. It's just that the `fma()` libcall in particular is problematic because:

1. For soft float, it's strictly more work than just doing the individual `fmul` + `fadd` compiler-rt libcalls.
2. It takes an extra dependency on libc even in freestanding.
3. It can cause the libc `fma()` implementation to call itself recursively (we observed this when compiling musl with `zig cc`).

Handling this special case in `isFMAFasterThanFMulAndFAdd()` seemed like the most straightforward solution, but I'm definitely open to suggestions.

**Comment by alexrp** [comment]
Updated tests to use IR `contract` and test vectors of `float`/`double`.

File: llvm/test/CodeGen/ARM/fmuladd-soft-float.ll [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+; SOFT-FLOAT-VFP64-NEXT:    push {r4, lr}
+; SOFT-FLOAT-VFP64-NEXT:    mov r4, r2
+; SOFT-FLOAT-VFP64-NEXT:    bl __mulsf3
+; SOFT-FLOAT-VFP64-NEXT:    mov r1, r4
+; SOFT-FLOAT-VFP64-NEXT:    bl __addsf3
+; SOFT-FLOAT-VFP64-NEXT:    pop {r4, lr}
+; SOFT-FLOAT-VFP64-NEXT:    mov pc, lr
+    %product = fmul contract float %a, %b
+    %result = fadd contract float %product, %c
+    ret float %result
```
Comment (arsenm) [inline llvm/test/CodeGen/ARM/fmuladd-soft-float.ll:106]: 2 space indent

**Review by arsenm — APPROVED** [review]
lgtm with nit

**Comment by alexrp** [comment]
ping (CI failure looks like unrelated breakage on main)

File: llvm/test/CodeGen/ARM/fmuladd-soft-float.ll (hunk 5 above)
Comment (alexrp) [reply llvm/test/CodeGen/ARM/fmuladd-soft-float.ll]: Fixed along with a rebase.

**Pull request merged**